staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go

```diff
@@ -396,47 +396,52 @@ func validateConvertedObject(in, out *unstructured.Unstructured) error {
 
 // restoreObjectMeta copies metadata from original into converted, while preserving labels and annotations from converted.
 func restoreObjectMeta(original, converted *unstructured.Unstructured) error {
-	obj, found := converted.Object["metadata"]
-	if !found {
+	cm, found := converted.Object["metadata"]
+	om, previouslyFound := original.Object["metadata"]
+	switch {
+	case !found && !previouslyFound:
+		return nil
+	case previouslyFound && !found:
 		return fmt.Errorf("missing metadata in converted object")
+	case !previouslyFound && found:
+		om = map[string]interface{}{}
 	}
-	responseMetaData, ok := obj.(map[string]interface{})
+
+	convertedMeta, ok := cm.(map[string]interface{})
 	if !ok {
-		return fmt.Errorf("invalid metadata of type %T in converted object", obj)
+		return fmt.Errorf("invalid metadata of type %T in converted object", cm)
 	}
-
-	if _, ok := original.Object["metadata"]; !ok {
-		// the original will always have metadata. But just to be safe, let's clear in converted
-		// with an empty object instead of nil, to be able to add labels and annotations below.
-		converted.Object["metadata"] = map[string]interface{}{}
-	} else {
-		converted.Object["metadata"] = original.Object["metadata"]
+	originalMeta, ok := om.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("invalid metadata of type %T in input object", om)
 	}
 
-	obj = converted.Object["metadata"]
-	convertedMetaData, ok := obj.(map[string]interface{})
-	if !ok {
-		return fmt.Errorf("invalid metadata of type %T in input object", obj)
+	result := converted
```

Comment:
Is it worth having a shallow copy here? Might make things more confusing below?

Reply:
done